relayer/naive-strategy.go

```diff
@@ -184,62 +208,126 @@ func (nrs *NaiveStrategy) RelayPacketsUnorderedChan(src, dst *Chain, sp *RelaySe
-func (*NaiveStrategy) RelayPacketsOrderedChan(src, dst *Chain, sp *RelaySequences, sh *SyncHeaders) error {
+func (nrs *NaiveStrategy) RelayPacketsOrderedChan(src, dst *Chain, sp *RelaySequences, sh *SyncHeaders) error {
+	relayMsgs, err := nrs.buildRelayMsgs(src, dst, sp, sh)
+	if err != nil {
+		return err
+	}
+
+	if len(relayMsgs) == 0 {
+		src.Log(fmt.Sprintf("- No packets to relay between [%s]port{%s} and [%s]port{%s}", src.ChainID, src.PathEnd.PortID, dst.ChainID, dst.PathEnd.PortID))
+		return nil
+	}
+
+	for _, relayMsg := range relayMsgs {
+		// TODO: increase the amount of gas as the number of messages increases
+		// notify the user of that
+		// XXX: should this just be a single log entry or one log per relay transaction?
+		if relayMsg.Send(src, dst); relayMsg.success {
+			if len(relayMsg.Dst) > 1 {
+				dst.logPacketsRelayed(src, len(relayMsg.Dst)-1)
+			}
+			if len(relayMsg.Src) > 1 {
+				src.logPacketsRelayed(dst, len(relayMsg.Src)-1)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (nrs *NaiveStrategy) buildRelayMsgs(src, dst *Chain, sp *RelaySequences, sh *SyncHeaders) ([]*RelayMsgs, error) {
+	var txSize int
 
 	// create the appropriate update client messages
 	msgs := &RelayMsgs{Src: []sdk.Msg{}, Dst: []sdk.Msg{}}
 
 	// add messages for src -> dst
-	for _, seq := range sp.Src {
+	for i, seq := range sp.Src {
 		chain, msg, err := packetMsgFromTxQuery(src, dst, sh, seq)
 		if err != nil {
-			return err
+			return nil, err
+		}
+
+		txSize += len(msg.GetSignBytes())
+
+		// enforce a maximum message length and maximum size of combined messages
+		if nrs.IsMaxTx(i, txSize) {
+			// update sequences processed
+			sp.Src = sp.Src[i:]
+			break
 		}
+
 		if chain == dst {
 			msgs.Dst = append(msgs.Dst, msg)
 		} else {
 			msgs.Src = append(msgs.Src, msg)
 		}
+
+		// clear sequence queue
+		if i == len(sp.Src)-1 {
+			sp.Src = []uint64{}
+		}
 	}
 
+	// reset txSize
+	txSize = 0
+
 	// add messages for dst -> src
-	for _, seq := range sp.Dst {
+	for i, seq := range sp.Dst {
 		chain, msg, err := packetMsgFromTxQuery(dst, src, sh, seq)
 		if err != nil {
-			return err
+			return nil, err
+		}
+
+		txSize += len(msg.GetSignBytes())
+
+		// enforce a maximum message length and maximum size of combined messages
+		if nrs.IsMaxTx(i, txSize) {
+			// update sequences processed
+			sp.Dst = sp.Dst[i:]
+			break
 		}
+
 		if chain == src {
 			msgs.Src = append(msgs.Src, msg)
 		} else {
 			msgs.Dst = append(msgs.Dst, msg)
 		}
-	}
 
-	if !msgs.Ready() {
-		src.Log(fmt.Sprintf("- No packets to relay between [%s]port{%s} and [%s]port{%s}", src.ChainID, src.PathEnd.PortID, dst.ChainID, dst.PathEnd.PortID))
-		return nil
+		// clear sequence queue
+		if i == len(sp.Dst)-1 {
+			sp.Dst = []uint64{}
+		}
 	}
 
 	// Prepend non-empty msg lists with UpdateClient
+	// XXX: should this be done just once since relay sequences aren't dynamically added after the call to this function?
```

Comment:
should an update client message be prepended before each relay tx or just once for the RelayPacketsOrderedChan?

Reply:
I think we can get away with just one? Cause the header will include all the proofs for the packets.